sdk/client/src/blocking.rs

```diff
@@ -0,0 +1,237 @@
+
+    //
+    // Private Helpers
+    //
+
+    fn send<T: DeserializeOwned>(&self, request: MethodRequest) -> Result<Response<T>> {
+        let request = JsonRpcRequest::new(request);
+        let resp: diem_json_rpc_types::response::JsonRpcResponse = self.send_impl(&request)?;
+
+        let (id, state, result) = validate(&resp)?;
+
+        if request.id() != id {
+            return Err(Error::rpc_response("invalid response id"));
+        }
+
+        let inner = serde_json::from_value(result).map_err(Error::decode)?;
+        Ok(Response::new(inner, state))
+    }
+
+    fn send_batch(
+        &self,
+        requests: Vec<MethodRequest>,
+    ) -> Result<Vec<Result<Response<MethodResponse>>>> {
+        let request: Vec<JsonRpcRequest> = requests.into_iter().map(JsonRpcRequest::new).collect();
+        let resp: BatchResponse = self.send_impl(&request)?;
+
+        let resp = resp.success()?;
+
+        validate_batch(&request, resp)
+    }
+
+    // Executes the specified request method using the given parameters by contacting the JSON RPC
+    // server. If the 'http_proxy' or 'https_proxy' environment variable is set, enable the proxy.
+    fn send_impl<S: Serialize, T: DeserializeOwned>(&self, payload: &S) -> Result<T> {
+        let mut request = ureq::post(&self.url)
+            .timeout_connect(REQUEST_TIMEOUT)
+            .build();
+
+        let proxy = proxy::Proxy::new();
+        let host = request.get_host().expect("unable to get the host");
+        let scheme = request
+            .get_scheme()
+            .expect("Unable to get the scheme from the host");
+        let proxy_url = match scheme.as_str() {
+            "http" => proxy.http(&host),
+            "https" => proxy.https(&host),
+            _ => None,
+        };
+        if let Some(proxy_url) = proxy_url {
+            request.set_proxy(ureq::Proxy::new(proxy_url).expect("Unable to parse proxy_url"));
+        }
+
+        let resp = request.send_json(serde_json::json!(payload));
+
+        if resp.synthetic() {
+            let e = resp.into_synthetic_error().unwrap();
+            let error = match &e {
+                ureq::Error::BadUrl(_)
+                | ureq::Error::UnknownScheme(_)
+                | ureq::Error::DnsFailed(_)
+                | ureq::Error::BadHeader
+                | ureq::Error::BadProxy
+                | ureq::Error::BadProxyCreds
+                | ureq::Error::ProxyConnect
+                | ureq::Error::InvalidProxyCreds
+                | ureq::Error::ConnectionFailed(_)
+                | ureq::Error::TlsError(_) => Error::request(e),
+                ureq::Error::Io(io_error) => {
+                    if let std::io::ErrorKind::TimedOut = io_error.kind() {
+                        Error::timeout(e)
+                    } else {
+                        Error::unknown(e)
+                    }
+                }
+                _ => Error::unknown(e),
+            };
+
+            return Err(error);
+        }
```

Comment:
Interesting, is this not on the non-blocking client?  I noticed that there's just generally different pieces in the `blocking` client around proxies and stuff.  I would have assumed it basically be a copy and paste + awaits and asyncs.

Reply:
Blocking and async clients use a completely different networking stack which means that most of its the same except for actually sending the network request